src/rdf4cpp/rdf/parser/RDFFileParser.hpp

```diff
@@ -0,0 +1,68 @@
+#ifndef RDF4CPP_RDFFILEPARSER_HPP
+#define RDF4CPP_RDFFILEPARSER_HPP
+
+#include "IStreamQuadIterator.hpp"
+#include <fstream>
+
+namespace rdf4cpp::rdf::parser {
+/**
+     * Similar to rdf4cpp::rdf::parser::IStreamQuadIterator
+     * Parses the file by the given path and tries to extract Quads given in TURTLE format.
+     *
+     * @note the iterator _starts on_ the first Quad
+     * @note An exhausted iterator becomes the end-of-stream iterator.
+     * @warning The value pointed to by an end-of-stream iterator is undefined
+     *
+     * @example
+     * @code
+     * for (const auto& e : rdf4cpp::rdf::parser::RDFFileParser{"somewhere.ttl"}) {
+     *  if (e.has_value())
+     *     std::cout << e.value();
+     *  else
+     *     std::cout << e.error();
+     * }
+     * @endcode
+     */
+class RDFFileParser {
+    std::string file_path_;
+    ParsingFlags flags_;
+    storage::node::NodeStorage node_storage_;
+
+public:
+    explicit RDFFileParser(const std::string &file_path, ParsingFlags flags = ParsingFlags::none(),
+                           storage::node::NodeStorage node_storage = storage::node::NodeStorage::default_instance());
+    explicit RDFFileParser(std::string &&file_path, ParsingFlags flags = ParsingFlags::none(),
+                           storage::node::NodeStorage node_storage = storage::node::NodeStorage::default_instance());
+
+    class Iterator {
+        friend class RDFFileParser;
+        friend bool operator==(const RDFFileParser::Iterator &iter, std::default_sentinel_t) noexcept;
+        std::ifstream stream_;
+        IStreamQuadIterator iter_;
```

Comment:
This here is actually UB; very hard to trigger UB. I couldn't get it to trigger in GCC at all for some reason, but using clang with `-O0 -fno-elide-constructors -fsanitize=address` the following code snippet crashes:

```c++
rdf4cpp::rdf::parser::RDFFileParser parser{"./tests_RDFFileParser_simple.ttl"};
auto it1 = parser.begin();
++it1;
```

**TLDR:**
The core problem is that moving a self-referential struct (which iterator is) is in almost all cases UB (except if you specifically account for that).

**Long version:**
The reason is that the iterator is moved once (because I forced the compiler to actually do it, normally this would be elided). Moving the iterator changes the address of the iteratator object, and therefore the address of `stream_` as well. But `IStreamQuadIterator` internally holds a pointer to the given stream, which of course still points to the old address of `stream_`, i.e. it points into garbage memory because `stream_` was moved elsewhere (into `it1` from the return slot of `begin()`).

Basically moving the iterator (and then using it) is UB at the moment, even if its hard to trigger UB because the optimizer is obscuring it. Thats why the snippet from the issue had `unique_ptrs`, because they heap allocate and thus effectively "pin" the object in place, the only thing you are moving then when moving the iterator is the pointer to the object, which is of course fine.